src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -765,10 +765,38 @@ else if (command instanceof MultiRangeReadCommand)
      *
      * @param iterator iterator over the keys from the index(es)
      */
-    float estimateSelectivity(RangeIterator iterator)
+    private float estimateSelectivity(RangeIterator iterator)
     {
-        float selectivity = Math.min((float) iterator.getMaxKeys() / estimateTotalAvailableRows(), 1.0f);
-        queryContext.setPostFilterSelectivityEstimate(selectivity);
+        long totalRowCount = estimateTotalAvailableRows();
+        return estimateSelectivity(iterator, totalRowCount);
+    }
+
+    private float estimateSelectivity(RangeIterator iterator, long totalRowCount)
+    {
+        float selectivity;
+
+        // For intersection and union we assume predicates are independent
+        // (column values are not correlated with each other).
+        // This assumption is not neccesarily true, but we have no statistical information to do any better.
+        if (iterator instanceof RangeIntersectionIterator)
+        {
+            RangeIntersectionIterator intersectionIterator = (RangeIntersectionIterator) iterator;
+            selectivity = 1.0f;
+            for (RangeIterator range : intersectionIterator.ranges)
```

Comment:
It's worth noting this is after we have applied the intersection clause limit. (I had to double check, so leaving a comment here.)